rpki-validator/src/main/java/net/ripe/rpki/validator3/api/bgp/BgpPreviewService.java

```diff
@@ -184,24 +189,33 @@ public int getEffectiveLength() {
     @lombok.Value
     public static class BgpPreviewEntry {
         // store it in memory optimised way, as we are going to store a lot of these objects in memory
-        long origin;
-        PackedIpRange prefix;
+        int origin;
+        short prefixLength; // negative for IPv6
+        long prefixHi;
+        long prefixLo;
         Validity validity;
 
-        BgpPreviewEntry(Asn origin, IpRange prefix, Validity validity) {
-            this.origin = origin.longValue();
-            this.prefix = new PackedIpRange(prefix);
-            this.validity = validity;
-        }
-
-        BgpPreviewEntry(long origin, PackedIpRange prefix, Validity validity) {
-            this.origin = origin;
-            this.prefix = prefix;
-            this.validity = validity;
-        }
-
         public static BgpPreviewEntry of(Asn origin, IpRange prefix, Validity validity) {
-            return new BgpPreviewEntry(origin, prefix, validity);
+            short prefixLength;
+            long prefixHi;
+            long prefixLo;
+            switch (prefix.getType()) {
+                case IPv4:
+                    prefixLength = (short) prefix.getPrefixLength();
+                    prefixHi = 0;
+                    prefixLo = ((Ipv4Address) prefix.getStart()).longValue();
+                    break;
+                case IPv6: {
+                    BigInteger value = prefix.getStart().getValue();
```

Comment:
No idea which approach is better, but here's another ipv6 stored as long implementation, and function that initialize it from BigInteger. Not quite sure why it goes first to byte buffer:
https://github.com/sargun/java-ipv6/blob/master/src/main/java/com/googlecode/ipv6/IPv6Address.java#L157

Reply:
I prefer shifting + conversion, since it more closely reflects the required math. Performance is probably very similar and not critical anyway.